pkg/controller/pxc/controller.go

```diff
@@ -534,6 +476,113 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
+		}
+
+		return nil
+	}
+
+	if err := r.enableHAProxy(cluster); err != nil {
+		return errors.Wrap(err, "HAProxy upgrade error")
+	}
+
+	if err := r.enableHAProxyService(cluster, owner); err != nil {
+		return errors.Wrap(err, "failed to create or update haproxy service")
+	}
+
+	if cluster.HAProxyReplicasServiceEnabled() {
+		if err := r.enableHAProxyReplicasService(cluster, owner); err != nil {
+			return errors.Wrap(err, "failed to create or update haproxy replicas")
+		}
+	} else {
+		if err := r.disableHAProxyReplicasService(cluster); err != nil {
+			return errors.Wrap(err, "delete HAProxy replica service")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) enableHAProxy(cluster *api.PerconaXtraDBCluster) error {
+	sfs := statefulset.NewHAProxy(cluster)
+	ps := &cluster.Spec.HAProxy.PodSpec
+
+	return r.updatePod(sfs, ps, cluster, nil)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) enableHAProxyService(cluster *api.PerconaXtraDBCluster, owner metav1.OwnerReference) error {
+	haProxyService := pxc.NewServiceHAProxy(cluster, owner)
+	haProxyService.Spec.Type = corev1.ServiceTypeClusterIP
+
+	if len(cluster.Spec.HAProxy.ServiceType) > 0 {
+		haProxyService.Spec.Type = cluster.Spec.HAProxy.ServiceType
+	}
+
+	if haProxyService.Spec.Type == corev1.ServiceTypeLoadBalancer ||
+		haProxyService.Spec.Type == corev1.ServiceTypeNodePort {
+		haProxyService.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
+
+		if len(cluster.Spec.HAProxy.ExternalTrafficPolicy) > 0 {
+			haProxyService.Spec.ExternalTrafficPolicy = cluster.Spec.HAProxy.ExternalTrafficPolicy
+		}
+	}
+
+	return r.createOrUpdate(haProxyService)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) enableHAProxyReplicasService(cluster *api.PerconaXtraDBCluster, owner metav1.OwnerReference) error {
+	haProxyServiceReplicas := pxc.NewServiceHAProxyReplicas(cluster, owner)
+
+	replicaPorts := []corev1.ServicePort{
+		{
+			Port:       3306,
+			TargetPort: intstr.FromInt(3307),
+			Name:       "mysql-replicas",
+		},
+	}
+
+	haProxyServiceReplicas.Spec.Ports = replicaPorts
+	haProxyServiceReplicas.Spec.Type = corev1.ServiceTypeClusterIP
+
+	if len(cluster.Spec.HAProxy.ReplicasServiceType) > 0 {
+		haProxyServiceReplicas.Spec.Type = cluster.Spec.HAProxy.ReplicasServiceType
+	}
+
+	if haProxyServiceReplicas.Spec.Type == corev1.ServiceTypeLoadBalancer ||
+		haProxyServiceReplicas.Spec.Type == corev1.ServiceTypeNodePort {
+		haProxyServiceReplicas.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
+
+		if len(cluster.Spec.HAProxy.ReplicasExternalTrafficPolicy) > 0 {
+			haProxyServiceReplicas.Spec.ExternalTrafficPolicy = cluster.Spec.HAProxy.ReplicasExternalTrafficPolicy
+		}
+	}
+
+	return r.createOrUpdate(haProxyServiceReplicas)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) disableHAProxy(o *api.PerconaXtraDBCluster) error {
+	return r.deleteStatefulSet(o, statefulset.NewHAProxy(o), false)
```

Comment:
I'm not sure about this functions.
They are again, just another function calls, without any actual "work"
@delgod what do you think?

Reply:
from one standpoint - the one-line function looks unnecessary.
from another standpoint - the name of the function logically describes what we are doing.
from a third standpoint - we can just add comments which describe calls, instead of creating one-line functions.

for this particular PR let's keep it as it is, BUT let's discuss together on Design call with Max and Ege how we should design such things